includes/models/model.llms.order.php

```diff
@@ -829,19 +868,54 @@ public function get_next_payment_due_date( $format = 'Y-m-d H:i:s' ) {
 	 *
 	 * @since 4.6.0
 	 *
-	 * @param string $action Action hook ID. Core actions are "llms_charge_recurring_payment" and "llms_access_plan_expiration".
+	 * @param string $action Action hook ID. Core actions are "llms_charge_recurring_payment", "llms_access_plan_expiration" and "llms_send_upcoming_payment_reminder_notification".
 	 * @return int|false Returns the timestamp of the next action as an integer or `false` when no action exist.
 	 */
 	public function get_next_scheduled_action_time( $action ) {
 		return as_next_scheduled_action( $action, $this->get_action_args() );
 	}
 
+	/**
+	 * Retrieve the date to remind user before actual payment
+	 *
+	 * @since [version]
+	 *
+	 * @param string|false $next_payment_date Optional. Next payment due date. If not provided it'll be retrieved using `$this->get_next_payment_due_date()`.
+	 * @param string       $format            Optional. Date return format. Default is 'Y-m-d H:i:s'.
+	 * @return WP_Error|string
+	 */
+	public function get_upcoming_payment_reminder_date( $next_payment_date = false, $format = 'Y-m-d H:i:s' ) {
+
+		$next_payment_date = false === $next_payment_date ? $this->get_next_payment_due_date() : $next_payment_date;
+
+		if ( ! $next_payment_date || is_wp_error( $next_payment_date ) ) {
+			return new WP_Error( 'plan-ended', __( 'No more payments due', 'lifterlms' ) );
+		}
+
+		/**
+		 * Filter the next upcoming payment reminder date.
+		 *
+		 * A timestamp should always be returned as the conversion to the requested format
+		 * will be performed on the returned value.
+		 *
+		 * @since [version]
+		 *
+		 * @param int        $upcoming_payment_reminder_time Unix timestamp for the next payment due date.
+		 * @param LLMS_Order $order                          Order object.
+		 * @param string     $format                         Requested date format.
+		 */
+		$upcoming_payment_reminder_time = apply_filters( 'llms_order_get_next_upcoming_payment_reminder_date', strtotime( '-1 day', strtotime( $next_payment_date ) ), $this, $format );
```

Comment:
I think it would make sense to add an additional filter allowing customization of the number of days directly. With this filter users will have to calculate the date themselves if they want to send the email 3 days before payment instead of 1.

We should leave this filter but I'm thinking just add another filter like `llms_order_payment_reminder_days`, default it to 1 and then users can quickly change 1 to 3 or 5 or 50.

This line can then pass in a variable (always converted to a negative number) instead of hardcoded 1.

Do you agree?

I also think that this is very likely something that many users will want to customize. 1 day maybe makes general sense but I actually think for me I'd want it to be 3 days.

I think we should add an option on the notification settings allowing end user customziation of this number without code

Reply:
I thought that too, and I have noted as something to ask you when the PR was ready.
Long story short: I fully agree.

Reply:
> I think we should add an option on the notification settings allowing end user customization of this number without code

On a second thought... this means that from the order model we're going to schedule an action which depends on a notification's setting?

Reply:
@thomasplevy ^

Reply:
Yes, that's what I'm thinking. The notification setting is something of a "global" setting that applies to all notifications of that type (regardless of order). I don't see a problem here. Do you see an issue that I'm not seeing or are you just clarifying?

Reply:
Well no it's fine, it's just that I already wasn't very sure that the method to retrieve the upcoming payment reminder of an order had to be a method of the order, but yeah it's fine so.